pkg/controllermanager/controller/shoot/shoot_maintenance_control.go

```diff
@@ -239,25 +291,31 @@ func hasMaintainNowAnnotation(shoot *gardencorev1beta1.Shoot) bool {
 }
 
 // MaintainMachineImages determines if a shoots machine images have to be maintained and in case returns the target images
-func MaintainMachineImages(shoot *gardencorev1beta1.Shoot, cloudProfile *gardencorev1beta1.CloudProfile, shootCurrentImages []*gardencorev1beta1.ShootMachineImage) ([]*gardencorev1beta1.ShootMachineImage, error) {
+func MaintainMachineImages(logger *logrus.Entry, shoot *gardencorev1beta1.Shoot, cloudProfile *gardencorev1beta1.CloudProfile) (updatedMachineImages []*gardencorev1beta1.ShootMachineImage, reasons []string, error error) {
 	var shootMachineImagesForUpdate []*gardencorev1beta1.ShootMachineImage
-	for _, shootImage := range shootCurrentImages {
-		machineImageFromCloudProfile, err := determineMachineImage(cloudProfile, shootImage)
+	var reasonsForUpdate []string
+	for _, worker := range shoot.Spec.Provider.Workers {
+		workerImage := worker.Machine.Image
+		machineImageFromCloudProfile, err := determineMachineImage(cloudProfile, workerImage)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 
-		shouldBeUpdated, shootMachineImage, err := shouldMachineImageBeUpdated(shoot, &machineImageFromCloudProfile, shootImage)
+		shouldBeUpdated, reason, updatedMachineImage, err := shouldMachineImageBeUpdated(logger, shoot.Spec.Maintenance.AutoUpdate.MachineImageVersion, &machineImageFromCloudProfile, workerImage)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 
-		if shouldBeUpdated {
-			shootMachineImagesForUpdate = append(shootMachineImagesForUpdate, shootMachineImage)
+		if !shouldBeUpdated {
+			return nil, nil, nil
 		}
-	}
 
-	return shootMachineImagesForUpdate, nil
+		message := fmt.Sprintf("image of worker-pool '%s' from version '(%s/%s)' to version '(%s/%s)'. Reason: %s", worker.Name, workerImage.Name, *workerImage.Version, updatedMachineImage.Name, *updatedMachineImage.Version, *reason)
```

Comment:
The image name will always stay the same, correct? Or am I mistaken?

Reply:
Yes correct, I just found it nice to prefix it with the machine image name - but I can change that.

Reply:
Sure, including the image name is good, I think. I just noticed that the message will include it twice :)

Reply:
ping